src/gui/editors/PianoRoll.cpp

```diff
@@ -2873,214 +2878,257 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
+		if (m_pianoKeysVisible >= NumKeys)
 		{
-			// draw it!
-			p.drawPixmap( PIANO_X, y - m_whiteKeySmallHeight, WHITE_KEY_WIDTH, m_whiteKeySmallHeight,
-							*s_whiteKeySmallPm );
-			// update y-pos
-			y -= m_whiteKeySmallHeight / 2;
-			// move first black key down (we didn't draw whole
-			// white key so black key needs to be lifted down)
-			// (default for first_white_key_height =
-			// m_whiteKeySmallHeight, so m_whiteKeySmallHeight/2
-			// is smaller)
-			first_white_key_height = m_whiteKeySmallHeight / 2;
+			// always increase m_notesEditHeight to compensate
+			m_pianoKeysVisible = NumKeys;
+			m_notesEditHeight = height() - (NumKeys * m_keyLineHeight) -
+				PR_TOP_MARGIN - PR_BOTTOM_MARGIN;
+			partialKeyVisible = 0;
 		}
-		// check whether to draw a big or a small white key
-		if( prKeyOrder[key % KeysPerOctave] == PR_WHITE_KEY_SMALL )
+		topKey = qBound(0, m_startKey + m_pianoKeysVisible - 1, NumKeys - 1);
+		topNote = topKey % KeysPerOctave;
+		// if resizing the note edit area
+		if (m_action != ActionResizeNoteEditArea && partialKeyVisible != 0)
 		{
-			// draw a small one while checking if it is pressed or not
-			if( hasValidPattern() && m_pattern->instrumentTrack()->pianoModel()->isKeyPressed( key ) )
-			{
-				p.drawPixmap(PIANO_X, y - m_whiteKeySmallHeight, WHITE_KEY_WIDTH, m_whiteKeySmallHeight,
-							*s_whiteKeySmallPressedPm);
-			}
-			else
-			{
-				p.drawPixmap(PIANO_X, y - m_whiteKeySmallHeight, WHITE_KEY_WIDTH, m_whiteKeySmallHeight,
-							*s_whiteKeySmallPm);
-			}
-			// update y-pos
-			y -= m_whiteKeySmallHeight;
+			// adding height always means we don't have to add keys
+			m_notesEditHeight += partialKeyVisible;
+		}
+		updatePositionLineHeight();
+		int x, q = quantization(), tick;
 
+		// draw vertical quantization lines
+		// If we're over 100% zoom, we allow all quantization level grids
+		if (m_zoomingModel.value() <= 3)
+		{
+			// we're under 100% zoom
+			// allow quantization grid up to 1/24 for triplets
+			if (q % 3 != 0 && q < 8) { q = 8; }
+			// allow quantization grid up to 1/32 for normal notes
+			else if (q < 6) { q = 6; }
+		}
+		auto xCoordOfTick = [=](int tick) {
+			return WHITE_KEY_WIDTH + (
+				(tick - m_currentPosition) * m_ppb / MidiTime::ticksPerBar()
+			);
+		};
+		p.setPen(lineColor());
+		for (tick = m_currentPosition - m_currentPosition % q,
+			x = xCoordOfTick(tick);
+			x <= width();
+			tick += q, x = xCoordOfTick(tick))
+		{
+			p.drawLine(x, keyAreaTop(), x, noteEditBottom());
 		}
-		else if( prKeyOrder[key % KeysPerOctave] == PR_WHITE_KEY_BIG )
+
+		// draw horizontal grid lines and piano notes
+		p.setClipRect(0, keyAreaTop(), width(), keyAreaBottom() - keyAreaTop());
+		int key_line_y = keyAreaTop();
+		int grid_line_y = keyAreaTop() + m_keyLineHeight - 1;
+		static const int yTextCorrections[KeysPerOctave] = {
+		/*   C      D      E   F      G      A      B   */
+		    -4, 0, -8, 0, -6, -4, 0, -8, 0, -8, 0, -6
+		};
```

Comment:
You can't use these constants anymore because of #5442.

Reply:
I switched to using the one from the PR, and all the labels are way off now. 😩

Reply:
I've managed to fix the offsets by doing a little different calculations, but with Y zoom I've found there are some draw bugs now when certain keys are the top key.